File: clients/python/src/rayito/_payload.py
"""Constructor del `runHookPayload` (ADR-004).

Es el único canal per-VM de `run-microvm`. Lleva el sha256 del access token
del agente, nunca el token: AWS puede registrar el payload en CloudTrail.

Contrato con `rayd` (ARCHITECTURE.md "Auth interna"): el access token es
`base64url(secret)` sin padding; `token_sha256 = sha256(secret).hex()`, es
decir, el hash de los **bytes decodificados**, no de la cadena. `rayd`
decodifica `x-access-token`, tolera padding y compara en tiempo constante.
"""

from __future__ import annotations

import base64
import binascii
import hashlib
import json
import secrets
from collections.abc import Mapping
from typing import Final

from rayito._limits import RUN_HOOK_PAYLOAD_MAX_CHARS
from rayito.exceptions import InvalidArgumentException

PAYLOAD_VERSION: Final = 1
DEFAULT_USER: Final = "user"
DEFAULT_WORKDIR: Final = "/home/user"
ACCESS_TOKEN_BYTES: Final = 32
CPU_TIME_LIMIT_MIN_SECONDS: Final = 1
CPU_TIME_LIMIT_MAX_SECONDS: Final = 28_800


def generate_access_token() -> str:
    return encode_access_token(secrets.token_bytes(ACCESS_TOKEN_BYTES))


def encode_access_token(secret: bytes) -> str:
    return base64.urlsafe_b64encode(secret).rstrip(b"=").decode("ascii")


def decode_access_token(access_token: str) -> bytes:
    """Los bytes que `rayd` hashea. Falla si el token no es base64url canónico
    (alfabeto estricto, bits sobrantes a cero), igual que el decoder de `rayd`."""
    stripped = access_token.rstrip("=")
    padded = stripped + "=" * (-len(stripped) % 4)
    try:
        secret = base64.b64decode(padded, altchars=b"-_", validate=True)
    except (binascii.Error, ValueError) as exc:
        raise InvalidArgumentException("access_token no es base64url") from exc
    if not secret or encode_access_token(secret) != stripped:
        raise InvalidArgumentException("access_token no es base64url")
    return secret


def validate_access_token(access_token: str) -> str:
    decode_access_token(access_token)
    return access_token


def access_token_sha256(access_token: str) -> str:
    return hashlib.sha256(decode_access_token(access_token)).hexdigest()


def build_run_hook_payload(
    *,
    access_token: str,
    envs: Mapping[str, str] | None = None,
    metadata: Mapping[str, str] | None = None,
    user: str = DEFAULT_USER,
    workdir: str = DEFAULT_WORKDIR,
    cpu_time_limit: int | None = None,
) -> str:
    """Serializa el payload y falla si supera el límite del modelo (4096 chars).

    `envs` y `metadata` se omiten del JSON cuando están vacíos para no gastar
    caracteres. `metadata` son etiquetas no secretas que `rayd` devuelve en
    `Health` (viajan como `envs`: CloudTrail puede registrar el payload).
    `cpu_time_limit` (segundos de CPU, no de pared, `1..=28800`) viaja como
    `limits: {"cpu_seconds": N}` sólo cuando se pasa: `rayd` lo aplica como
    `RLIMIT_CPU` a cada proceso y PTY del sandbox, nunca al kernel.
    """
    if not access_token:
        raise InvalidArgumentException("access_token no puede estar vacío")
    payload: dict[str, object] = {
        "v": PAYLOAD_VERSION,
        "token_sha256": access_token_sha256(access_token),
        "user": user,
        "workdir": workdir,
    }
    if envs:
        payload["envs"] = validated_envs(envs)
    if metadata:
        payload["metadata"] = validated_metadata(metadata)
    if cpu_time_limit is not None:
        payload["limits"] = {"cpu_seconds": validated_cpu_time_limit(cpu_time_limit)}
    text = json.dumps(payload, separators=(",", ":"), sort_keys=True)
    if len(text) > RUN_HOOK_PAYLOAD_MAX_CHARS:
        raise InvalidArgumentException(
            f"runHookPayload ocupa {len(text)} caracteres y el máximo es "
            f"{RUN_HOOK_PAYLOAD_MAX_CHARS}. Reduce `envs` o `metadata` en create(); las "
            "variables grandes van en `envs` de cada comando o en un fichero con "
            "files.write()."
        )
    return text


def validated_cpu_time_limit(cpu_time_limit: int) -> int:
    if (
        isinstance(cpu_time_limit, bool)
        or not isinstance(cpu_time_limit, int)
        or not CPU_TIME_LIMIT_MIN_SECONDS <= cpu_time_limit <= CPU_TIME_LIMIT_MAX_SECONDS
    ):
        raise InvalidArgumentException(
            f"cpu_time_limit debe ser un entero entre {CPU_TIME_LIMIT_MIN_SECONDS} y "
            f"{CPU_TIME_LIMIT_MAX_SECONDS} segundos de CPU, recibido {cpu_time_limit!r}"
        )
    return cpu_time_limit


def validated_envs(envs: Mapping[str, str]) -> dict[str, str]:
    return validated_string_map(envs, field="envs")


def validated_metadata(metadata: Mapping[str, str]) -> dict[str, str]:
    return validated_string_map(metadata, field="metadata")


def validated_string_map(values: Mapping[str, str], *, field: str) -> dict[str, str]:
    for key, value in values.items():
        if not isinstance(key, str) or not key:
            raise InvalidArgumentException(f"clave de {field} inválida: {key!r}")
        if not isinstance(value, str):
            raise InvalidArgumentException(f"el valor de {field}[{key!r}] debe ser str")
    return dict(values)

File: clients/typescript/src/payload.ts
/**
 * Constructor del `runHookPayload` (ADR-004).
 *
 * Es el único canal per-VM de `run-microvm`. Lleva el sha256 del access token
 * del agente, nunca el token: AWS puede registrar el payload en CloudTrail.
 *
 * Contrato con `rayd` (ARCHITECTURE.md "Auth interna"): el access token es
 * `base64url(secret)` sin padding; `token_sha256 = sha256(secret).hex()`, es
 * decir, el hash de los **bytes decodificados**, no de la cadena. `rayd`
 * decodifica `x-access-token`, tolera padding y compara en tiempo constante.
 */

import { createHash, randomBytes } from "node:crypto";
import { InvalidArgumentError } from "./errors.js";
import { RUN_HOOK_PAYLOAD_MAX_CHARS } from "./limits.js";
import { type LifecycleBlock, lifecycleBlockToWire } from "./sandbox/lifecycle.js";

export const PAYLOAD_VERSION = 1;
export const DEFAULT_USER = "user";
export const DEFAULT_WORKDIR = "/home/user";
export const ACCESS_TOKEN_BYTES = 32;
export const CPU_TIME_LIMIT_MIN_SECONDS = 1;
export const CPU_TIME_LIMIT_MAX_SECONDS = 28_800;

const BASE64URL_ALPHABET = /^[A-Za-z0-9_-]+$/;

export function generateAccessToken(): string {
  return encodeAccessToken(randomBytes(ACCESS_TOKEN_BYTES));
}

export function encodeAccessToken(secret: Uint8Array): string {
  return Buffer.from(secret).toString("base64url");
}

/**
 * Los bytes que `rayd` hashea. Falla si el token no es base64url canónico
 * (alfabeto estricto, bits sobrantes a cero), igual que el decoder de `rayd`.
 */
export function decodeAccessToken(accessToken: string): Uint8Array {
  const stripped = accessToken.replace(/=+$/, "");
  if (!BASE64URL_ALPHABET.test(stripped)) {
    throw new InvalidArgumentError("accessToken no es base64url");
  }
  const secret = Buffer.from(stripped, "base64url");
  if (secret.length === 0 || encodeAccessToken(secret) !== stripped) {
    throw new InvalidArgumentError("accessToken no es base64url");
  }
  return new Uint8Array(secret);
}

export function validateAccessToken(accessToken: string): string {
  decodeAccessToken(accessToken);
  return accessToken;
}

export function accessTokenSha256(accessToken: string): string {
  return createHash("sha256").update(decodeAccessToken(accessToken)).digest("hex");
}

export interface RunHookPayloadOptions {
  readonly accessToken: string;
  readonly envs?: Readonly<Record<string, string>> | undefined;
  readonly metadata?: Readonly<Record<string, string>> | undefined;
  readonly cpuTimeLimit?: number | undefined;
  readonly user?: string | undefined;
  readonly workdir?: string | undefined;
  /**
   * `"network": {"enforce": true}` sólo cuando es `true`: `rayd` instala
   * deny-all antes de responder a `/run`. Nunca lleva reglas ni credenciales
   * (viajan después por `UpdateNetwork`, autenticado con el token).
   */
  readonly networkEnforce?: boolean | undefined;
  /** El plazo lógico (ADR-011), ya validado por `resolveLifecycle`; sin él no viaja la clave. */
  readonly lifecycle?: LifecycleBlock | undefined;
}

/**
 * Serializa el payload (JSON compacto, claves ordenadas, sólo ASCII como
 * `json.dumps` en Python) y falla si supera el límite del modelo (4096 chars).
 * `envs` y `metadata` se omiten del JSON cuando están vacíos para no gastar
 * caracteres; `cpuTimeLimit` (segundos de CPU, `1..=28800`) viaja como
 * `limits: {"cpu_seconds": N}` sólo cuando se pasa, igual que en Python.
 * `lifecycle` (≈ 80 caracteres) sólo viaja cuando se pidió un plazo lógico:
 * un `rayd` anterior a M9 ignora la clave y `v` sigue siendo 1.
 */
export function buildRunHookPayload(options: RunHookPayloadOptions): string {
  if (!options.accessToken) {
    throw new InvalidArgumentError("accessToken no puede estar vacío");
  }
  const payload: Record<string, unknown> = {
    v: PAYLOAD_VERSION,
    token_sha256: accessTokenSha256(options.accessToken),
    user: options.user ?? DEFAULT_USER,
    workdir: options.workdir ?? DEFAULT_WORKDIR,
  };
  if (options.envs !== undefined && Object.keys(options.envs).length > 0) {
    payload.envs = validatedEnvs(options.envs);
  }
  if (options.metadata !== undefined && Object.keys(options.metadata).length > 0) {
    payload.metadata = validatedStringMap(options.metadata, "metadata");
  }
  if (options.cpuTimeLimit !== undefined) {
    payload.limits = { cpu_seconds: validatedCpuTimeLimit(options.cpuTimeLimit) };
  }
  if (options.networkEnforce === true) {
    payload.network = { enforce: true };
  }
  if (options.lifecycle !== undefined) {
    payload.lifecycle = lifecycleBlockToWire(options.lifecycle);
  }
  const text = asciiJson(sortedKeys(payload));
  if (text.length > RUN_HOOK_PAYLOAD_MAX_CHARS) {
    throw new InvalidArgumentError(
      `runHookPayload ocupa ${text.length} caracteres y el máximo es ` +
        `${RUN_HOOK_PAYLOAD_MAX_CHARS}. Reduce \`envs\` en create(); las variables ` +
        "grandes van en `envs` de cada comando o en un fichero con files.write().",
    );
  }
  return text;
}

export function validatedEnvs(envs: Readonly<Record<string, string>>): Record<string, string> {
  return validatedStringMap(envs, "envs");
}

export function validatedStringMap(
  values: Readonly<Record<string, string>>,
  field: string,
): Record<string, string> {
  const validated: Record<string, string> = {};
  for (const [key, value] of Object.entries(values)) {
    if (typeof key !== "string" || key.length === 0) {
      throw new InvalidArgumentError(`clave de ${field} inválida: ${JSON.stringify(key)}`);
    }
    if (typeof value !== "string") {
      throw new InvalidArgumentError(
        `el valor de ${field}[${JSON.stringify(key)}] debe ser string`,
      );
    }
    validated[key] = value;
  }
  return validated;
}

export function validatedCpuTimeLimit(cpuTimeLimit: unknown): number {
  if (
    typeof cpuTimeLimit !== "number" ||
    !Number.isInteger(cpuTimeLimit) ||
    cpuTimeLimit < CPU_TIME_LIMIT_MIN_SECONDS ||
    cpuTimeLimit > CPU_TIME_LIMIT_MAX_SECONDS
  ) {
    throw new InvalidArgumentError(
      `cpuTimeLimit debe ser un entero entre ${CPU_TIME_LIMIT_MIN_SECONDS} y ` +
        `${CPU_TIME_LIMIT_MAX_SECONDS} segundos de CPU, recibido ${String(cpuTimeLimit)}`,
    );
  }
  return cpuTimeLimit;
}

function sortedKeys(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortedKeys);
  }
  if (typeof value === "object" && value !== null) {
    const entries = Object.entries(value as Record<string, unknown>).sort(([a], [b]) =>
      a < b ? -1 : a > b ? 1 : 0,
    );
    return Object.fromEntries(entries.map(([key, item]) => [key, sortedKeys(item)]));
  }
  return value;
}

function asciiJson(value: unknown): string {
  return JSON.stringify(value).replace(
    /[-￿]/g,
    (char) => `\\u${char.charCodeAt(0).toString(16).padStart(4, "0")}`,
  );
}
